server/src/test/java/org/elasticsearch/common/xcontent/support/XContentMapValuesTests.java

```diff
@@ -730,4 +732,175 @@ public void testExtractNestedSources() {
+            builder.endObject();
+
+            Map<String, Object> map = toSourceMap(Strings.toString(builder));
+            XContentMapValues.insertValue("path1.path2.test", map, "value2");
+            assertThat(getMapValue.apply(map, "path1.path2.test"), equalTo("value2"));
+            XContentMapValues.insertValue("path1.path2.test_me", map, "test_me_value");
+            assertThat(getMapValue.apply(map, "path1\\.path2\\.test_me"), equalTo("test_me_value"));
+            XContentMapValues.insertValue("path1.non_path2.test", map, "test_value");
+            assertThat(getMapValue.apply(map, "path1\\.non_path2\\.test"), equalTo("test_value"));
+
+            IllegalArgumentException ex = assertThrows(
+                IllegalArgumentException.class,
+                () -> XContentMapValues.insertValue("path1.path2.test.test2", map, "value3")
+            );
+            assertThat(
+                ex.getMessage(),
+                equalTo("Path [path1.path2.test] has value [value2] of type [String], which cannot be traversed into further")
+            );
+
+            XContentMapValues.insertValue("path1.path2", map, Map.of("path3", "bar"));
+            assertThat(getMapValue.apply(map, "path1.path2"), equalTo(Map.of("path3", "bar")));
+
+            XContentMapValues.insertValue("path1", map, "baz");
+            assertThat(getMapValue.apply(map, "path1"), equalTo("baz"));
+        }
+
+        // lists
+        {
+            XContentBuilder builder = XContentFactory.jsonBuilder().startObject();
+            builder.startObject("path1").array("test", "value1", "value2").endObject();
+            builder.endObject();
+            Map<String, Object> map = toSourceMap(Strings.toString(builder));
+
+            XContentMapValues.insertValue("path1.test", map, List.of("value3", "value4", "value5"));
+            assertThat(getMapValue.apply(map, "path1.test"), equalTo(List.of("value3", "value4", "value5")));
+        }
+        {
+            XContentBuilder builder = XContentFactory.jsonBuilder().startObject();
+            {
+                builder.startObject("path1");
+                {
+                    builder.startArray("path2");
+                    builder.startObject().field("test1", "value1").endObject();
+                    builder.startObject().field("test2", "value2").endObject();
+                    builder.startObject().field("test2", "value3").endObject();
+                    builder.endArray();
+                }
+                builder.endObject();
+                builder.startObject("path1.path2");
+                {
+                    builder.array("test2", "value3", "value4");
+                }
+                builder.endObject();
+            }
+            builder.endObject();
+            Map<String, Object> map = toSourceMap(Strings.toString(builder));
+
+            XContentMapValues.insertValue("path1.path2.test1", map, "new_value");
```

Comment:
That's the case that I think should throw an error. In this case I count 3 possible places for `test1` to be added.
That's ambiguous since if we wanted to put the original value back we'd need to store which index it appears in the `path2` array of objects.

Reply:
I don't personally see how this is ambiguous, there is only one way to traverse to the existing value of `path1.path2.test1`. If we wanted to revert the change and put the original value back, the logic in `extractSuffixMaps` would allow us to do that.

If, however, we want to handle lists of maps in general as an ambiguous layout (whether replacing an existing value or adding a new one), that's fine, but it feels like an artificial limitation to me.

Reply:
> here is only one way to traverse to the existing value of path1.path2.test1

This is true if we keep `test1` in source but I am anticipating that we'll want to remove the semantic text object. 
Once removed, we have no way to know which object in the list should have the original value. 
That's an implementation detail for us but something I'd like to keep simple. If we start accepting all weird json layout we'll make the rest of the work more difficult to achieve imo.